pkg/controller.v1/pytorch/pytorchjob_controller_test.go

```diff
@@ -164,32 +192,335 @@ var _ = Describe("PyTorchJob controller", func() {
 			// Check if the job is succeeded.
 			cond := getCondition(created.Status, kubeflowv1.JobSucceeded)
 			Expect(cond.Status).To(Equal(corev1.ConditionTrue))
-			By("Deleting the PyTorchJob")
-			Expect(testK8sClient.Delete(ctx, job)).Should(Succeed())
+		})
+
+		It("Shouldn't create resources if PyTorchJob is suspended; Should create resources once PyTorchJob is unsuspended", func() {
+			By("By creating a new PyTorchJob with suspend=true")
+			job.Spec.RunPolicy.Suspend = pointer.Bool(true)
+			job.Spec.PyTorchReplicaSpecs[kubeflowv1.PyTorchJobReplicaTypeWorker].Replicas = pointer.Int32(1)
+			Expect(testK8sClient.Create(ctx, job)).Should(Succeed())
+
+			created := &kubeflowv1.PyTorchJob{}
+			masterPod := &corev1.Pod{}
+			workerPod := &corev1.Pod{}
+			masterSvc := &corev1.Service{}
+			workerSvc := &corev1.Service{}
+
+			By("Checking created PyTorchJob")
+			Eventually(func() bool {
+				err := testK8sClient.Get(ctx, jobKey, created)
+				return err == nil
+			}, testutil.Timeout, testutil.Interval).Should(BeTrue())
+			By("Checking created PyTorchJob has a nil startTime")
+			Consistently(func() *metav1.Time {
+				Expect(testK8sClient.Get(ctx, jobKey, created)).Should(Succeed())
+				return created.Status.StartTime
+			}, testutil.ConsistentDuration, testutil.Interval).Should(BeNil())
+
+			By("Checking if the pods and services aren't created")
+			Consistently(func() bool {
+				errMasterPod := testK8sClient.Get(ctx, masterKey, masterPod)
+				errWorkerPod := testK8sClient.Get(ctx, worker0Key, workerPod)
+				errMasterSvc := testK8sClient.Get(ctx, masterKey, masterSvc)
+				errWorkerSvc := testK8sClient.Get(ctx, worker0Key, workerSvc)
+				return errors.IsNotFound(errMasterPod) && errors.IsNotFound(errWorkerPod) &&
+					errors.IsNotFound(errMasterSvc) && errors.IsNotFound(errWorkerSvc)
+			}, testutil.ConsistentDuration, testutil.Interval).Should(BeTrue())
+
+			By("Checking if the PyTorchJob has suspended condition")
+			Eventually(func() []kubeflowv1.JobCondition {
+				Expect(testK8sClient.Get(ctx, jobKey, created)).Should(Succeed())
+				return created.Status.Conditions
+			}, testutil.ConsistentDuration, testutil.Interval).Should(BeComparableTo([]kubeflowv1.JobCondition{
+				{
+					Type:    kubeflowv1.JobCreated,
+					Status:  corev1.ConditionTrue,
+					Reason:  commonutil.NewReason(kubeflowv1.PytorchJobKind, commonutil.JobCreatedReason),
+					Message: fmt.Sprintf("PyTorchJob %s is created.", name),
+				},
+				{
+					Type:    kubeflowv1.JobSuspended,
+					Status:  corev1.ConditionTrue,
+					Reason:  commonutil.NewReason(kubeflowv1.PytorchJobKind, commonutil.JobSuspendedReason),
+					Message: fmt.Sprintf("PyTorchJob %s is suspended.", name),
+				},
+			}, testutil.IgnoreJobConditionsTimes))
+
+			By("Unsuspending the PyTorchJob")
```

Comment:
While the part of the test case about creating a suspended job is only tested here, the unsuspending appears to be covered already in the test case below. Consider if it adds value here.

Reply:
It makes sense. That is duplicated case. I will remove the unsuspending case from here.

Reply:
Done.